pylint/checkers/typecheck.py

```diff
@@ -1951,6 +1956,18 @@ def visit_generatorexp(self, node):
         for gen in node.generators:
             self._check_iterable(gen.iter, check_async=gen.is_async)
 
+    @check_messages("await-outside-async")
+    def visit_await(self, node: astroid.Await) -> None:
+        self._check_await_outside_coroutine(node)
+
+    def _check_await_outside_coroutine(self, node: astroid.Await) -> None:
+        node_scope = node.scope()
+        while not isinstance(node_scope, astroid.Module):
+            if isinstance(node_scope, astroid.AsyncFunctionDef):
+                return
+            node_scope = node_scope.parent.scope()
```

Comment:
That is better but doesn't quite work for nested functions:
```py
import asyncio

async def func():
    def inner_func():
        await asyncio.sleep(1)  # should be an error

    inner_func()

asyncio.run(func())
```

It might work if you do `break` on `astroid.FunctionDef`.

--
Would be great if you can add this to the test cases, as well as the other way round (async function inside a normal one).

Reply:
sounds great, nested async (with await) inside normal function should be fine right?

Reply:
> sounds great, nested async (with await) inside normal function should be fine right?

Yes, that's fine. It's the same as with the normal scope that's also NOT async.

Reply:
added new test cases in dda6bda6